`frozen-lockfile` is used on CI even if I didn’t ask for it

<!--
  If this issue affects many people in a company/big team, create a post for your company in the following discussion:
  https://github.com/pnpm/pnpm/discussions/3787
  and link the issue in your post.

  This will help us prioritize issues that affect more people.
-->

### pnpm version

6.24.2

### Code to reproduce the issue

https://github.com/browserslist/caniuse-lite/runs/4598871429

1. Script on CI changed `devDependencies`
2. CI called `pnpm install --ignore-scripts`
3. There was no `--frozen-lockfile`, but we still get

```
lockfile is up-to-date, resolution step is skipped
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with package.json
```

<!--
If there was a fatal error also include a gist of your node_modules/.pnpm-debug.log file.
-->

### Expected behavior:

1. Have `--no-frozen-lockfile` if `frozen-lockfile` is default CI behavior by default
2. Or change lockfile if it was not expected behavior

### Actual behavior:

` ERR_PNPM_OUTDATED_LOCKFILE ` error without argument to disable it.

(We can’t use `pnpm up` instead of `pnpm install` because we need `--ignore-scripts` for security reasons)

### Additional information:

 - `node -v` prints: 16.13.1
 - Windows, macOS, or Linux?: Linux

--- a/.changeset/fair-hotels-suffer.md
+++ b/.changeset/fair-hotels-suffer.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/core": patch
+"@pnpm/headless": patch
+---
+
+Add more details to the frozen lockfile error.

--- a/.changeset/many-colts-think.md
+++ b/.changeset/many-colts-think.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/plugin-commands-installation": patch
+---
+
+Document `--no-frozen-lockfile` explicitly in the help of the install command.

--- a/packages/core/src/install/index.ts
+++ b/packages/core/src/install/index.ts
@@ -251,7 +251,9 @@ export async function mutateModules (
       )
     ) {
       if (needsFullResolution) {
-        throw new PnpmError('FROZEN_LOCKFILE_WITH_OUTDATED_LOCKFILE', 'Cannot perform a frozen installation because the lockfile needs updates')
+        throw new PnpmError('FROZEN_LOCKFILE_WITH_OUTDATED_LOCKFILE', 'Cannot perform a frozen installation because the lockfile needs updates', {
+          hint: 'Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"',
+        })
       }
       if (!ctx.existsWantedLockfile) {
         if (ctx.projects.some((project) => pkgHasDependencies(project.manifest))) {

--- a/packages/headless/src/index.ts
+++ b/packages/headless/src/index.ts
@@ -155,7 +155,9 @@ export default async (opts: HeadlessOptions) => {
       if (!satisfiesPackageManifest(wantedLockfile, manifest, id)) {
         throw new PnpmError('OUTDATED_LOCKFILE',
           `Cannot install with "frozen-lockfile" because ${WANTED_LOCKFILE} is not up-to-date with ` +
-          path.relative(lockfileDir, path.join(rootDir, 'package.json')))
+          path.relative(lockfileDir, path.join(rootDir, 'package.json')), {
+            hint: 'Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"',
+          })
       }
     }
   }

--- a/packages/plugin-commands-installation/src/install.ts
+++ b/packages/plugin-commands-installation/src/install.ts
@@ -123,8 +123,8 @@ For options that may be used with `-r`, see "pnpm help recursive"',
             name: '--lockfile-only',
           },
           {
-            description: "Don't generate a lockfile and fail if an update is needed",
-            name: '--frozen-lockfile',
+            description: "Don't generate a lockfile and fail if an update is needed. This setting is on by default in CI environments, so use --no-frozen-lockfile if you need to disable it for some reason",
+            name: '--[no-]frozen-lockfile',
           },
           {
             description: `If the available \`${WANTED_LOCKFILE}\` satisfies the \`package.json\` then perform a headless installation`,

